### ISSUE TITLE
Generics and Inheritance

### ISSUE THREAD

**Issue by jontavakoli** [dotnet/aspnetcore#29105]
Consider the following code in C#, it executes just fine. As the middle tier class instantiates the Storage class as an Array of TValue.

```	
       public class Storage<TValue> {
		public TValue Values { get; set; }
	}

	public class BaseClass<TValue> : Storage<TValue[]> { }

	public class FinalClass<TValue> {

		private BaseClass<string> baseClass;

		public FinalClass() => this.baseClass = new BaseClass<string> { Values = new string[] { } }; //Important Line of Code
	}
```

Because the middle tier class defines the TValue parameter as string, it allows to set the Values property in the FinalClass as new string[].
However, this behavior breaks in Blazor, because of the RuntimeHelpers.TypeCheck method.

Storage.razor

```
@typeparam TValue

@code {
	
	[Parameter]
	public TValue Values { get; set; }

}
```

BaseClass.razor
```
@typeparam TValue
@inherits Storage<TValue[]>
```

FinalClass.razor
```
<BaseClass TValue="string" Values="@(new string[]{})"></BaseClass>
```

The error you receive is "cannot convert from 'string[]' to 'TValue[]'"

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us.
Components are expected to inherit from `ComponentBase` and you're trying to inherit from an arbitrary class here.
Can you update your code and see whether that will work?
If it doesn't work, please share a small repro app hosted in GitHub so we can look into it.

**Comment by jontavakoli** [comment]
The code I provide was simply to demonstrate the logics and inheritance, I have crated a sample project that will cause this issue.
[TestBlazorApp](https://github.com/jontavakoli/TestBlazorApp)

Under the Pages folder there are 4 important classes.

1. StorageBase.cs
```
using Microsoft.AspNetCore.Components;

namespace TestBlazorApp.Pages
{

    public class StorageBase<TItem> : ComponentBase
    {

        [Parameter]
        public TItem Items { get; set; }

        [Parameter]
        public EventCallback<TItem> ItemsChanged { get; set; }

    }

}
```

2. FinalBase.cs inherits from StorageBase altering the generic parameter definition.
```
namespace TestBlazorApp.Pages
{
    public class FinalBase<TItem> : StorageBase<TItem[]>
    {
    }
}
```

3. FinalComponent.razor a super simple component that inherits from FinalBase.
```
@typeparam TItem
@inherits FinalBase<TItem>

<div>Items</div>
@if (Items != null)
{
	@foreach (var item in Items)
	{
		<div>@item</div>
	}
}
```

4. Index.razor which utilizes the FinalComponent.razor
`<FinalComponent TItem="string" @bind-Items="model.Names"></FinalComponent>`

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us.
This is not something we plan to do in the near future. You can learn more about our triage process and how we handle issues by reading our [Triage Process](https://github.com/dotnet/aspnetcore/blob/master/docs/TriageProcess.md) writeup.

**Comment by SteveSandersonMS** [comment]
The repro for this can be simplified a lot, so it looks worse to me now. All you need is the following component:

```cs
    public class SomeComponent<T> : ComponentBase
    {
        [Parameter] public T[] Items { get; set; }
    }
```

You just can't use this component. If you don't specify the `T` parameter explicitly, the compiler complains that it can't infer the type and you should specify it. But if you do specify it, like this:

```razor
<SomeComponent T="string" Items="@(Array.Empty<string>())"></SomeComponent>
```

... then you get the same compiler error reported above: `Argument 1: cannot convert from 'string[]' to 'T[]'`. The generated code looks like this:

```cs
__builder.AddAttribute(
    1,
    "Items",
    global::Microsoft.AspNetCore.Components.CompilerServices.RuntimeHelpers.TypeCheck<T[]>(
        Array.Empty<string>()));
```

... which is wrong because the compiler hasn't substituted in the known value for `T`.

I don't know whether this would be a trivial fix or is just impossible given the design of the type checker, but it's definitely incorrect, so reopening.

**Comment by SteveSandersonMS** [comment]
Hopefully fixed in https://github.com/dotnet/aspnetcore/pull/34734 which we'll consider for 6.0

**Comment by mkArtakMSFT** [comment]
This should be resolved already by #34734
